
# Which TFs are upregulated in the treatment group?
?results

rna_seq_control_treatment <- results(rnaDDS, c("Group", "control", "treated"), format = "GRanges")

rna_seq_control_treatment <- rna_seq_control_treatment[order(rna_seq_control_treatment$pvalue)]
rna_seq_control_treatment

rnaDDS
rna_seq_control_treatment_2 <- results(rnaDDS, c("Group", "treated", "control"), format = "GRanges")
rna_seq_control_treatment_2 <- rna_seq_control_treatment_2[order(rna_seq_control_treatment_2$pvalue)]
rna_seq_control_treatment_2

toOverLap_genes <- genes(TxDb.Hsapiens.UCSC.hg38.knownGene, columns = "gene_id", filter=NULL, single.strand.genes.only=TRUE)
toOverLap_genes


# subsetting regions of significantly enriched open chromatin (in a 1000 bp distance from a promotor)
rna_seq_control_treatment_2_significant <- rna_seq_control_treatment_2[(!is.na(rna_seq_control_treatment_2$padj) & 
                                                                          rna_seq_control_treatment_2$padj < 0.01) & rna_seq_control_treatment_2 %over% toOverLap_genes, ]

rna_seq_control_treatment_2_significant
# 5077
makebedtable(rna_seq_control_treatment_2_significant, "rna_seq_control_treatment_2_significant.html", basedirectory = getwd())

save(rna_seq_control_treatment_2_significant,file="significant_deseq2_rna-seq_peaks(treatment_vs_control).Rda")

# annotating these peaks
anno_rna_peaks_tc <- annotatePeak(rna_seq_control_treatment_2_significant, tssRegion=c(-3000, 3000), TxDb = TxDb.Hsapiens.UCSC.hg38.knownGene)
anno_rna_peaks_tc

anno_rna_peaks_tc_df <- as.data.frame(anno_rna_peaks_tc)

################################################################

# Q3 Identifying candidate transcription factors that 
# may be responsible for the underlying regulation.

BiocManager::install("TFEA.ChIP")
library(TFEA.ChIP)

# the object anno_rna_peaks_df contains only annotated genes, 
# that were differntially expressed beyond a significance 
# threshold of 0.01

anno_rna_peaks_df_TF <- anno_rna_peaks_tc_df%>%
  mutate(Genes = anno_rna_peaks_tc_df$geneId)

table_TF <- preprocessInputData( anno_rna_peaks_df_TF )
table_TF

head(table_TF)

Genes.Upreg <- Select_genes( table_TF, min_LFC = 1 )
Genes.Upreg
length(Genes.Upreg)
# 736

Genes.Control <- Select_genes( table_TF,
                               min_LFC = -0.25, max_LFC = 0.25 )
Genes.Control

length(Genes.Control)
# 588

# GeneID2entrez( gene.IDs = c("EGLN3","NFYA","ALS2","MYC","ARNT" ) )
# "112399" "4800"   "57679"  "4609"   "405"  

# CM_list_UP <- contingency_matrix( Genes.Upreg, Genes.Control ) #generates list of contingency tables, one per dataset
# CM_list_UP
# type(CM_list_UP)
# list

# We apply Fisher's exact test to each contingency table to test 
# the Ho that TF binding and differential expression are independent.
# The function returns the raw P-values and the FDR-adjusted v-values

# pval_mat_UP <- getCMstats( CM_list_UP ) #generates list of p-values and OR from association test
# head( pval_mat_UP)
#    
# pval_mat_UP_df <- as.data.frame(pval_mat_UP) 
# dim(pval_mat_UP_df)
# # 1060   10
# significant_TFs_subsetted <- subset(pval_mat_UP_df, pval_mat_UP_df$adj.p.value < 0.01)
# enriched_TFs <- significant_TFs_subsetted$TF
# 
# summary(enriched_TFs)
# enriched_TFs_fct <- as.factor(enriched_TFs)
# 
# distinct_enriched_TFs <- levels(enriched_TFs_fct)
# length(distinct_enriched_TFs)
# 247
# You could use a different P-val threshold
# or use a different subsetted dataset 

# generate an index of the tables on interest 
index <- get_chip_index( encodeFilter = TRUE ) # Or select ENCODE datasets only
CM_list_UPe <- contingency_matrix( Genes.Upreg, Genes.Control, index ) #generates list of contingency tables

# supply the index to the getCMstats function
# we restrict the analysis to the datasets generated by the ENCODE project and use 
# all non-differentially expressed genes as a control
pval_mat_UPe <- getCMstats( CM_list_UPe, index ) #generates list of p-values and ORs
pval_mat_UPe



length(pval_mat_UPe)
length(pval_mat_UP)
# both length 10

pval_mat_UPe_df <- as.data.frame(pval_mat_UPe)
dim(pval_mat_UPe_df)
# 1060   10

save(pval_mat_UPe_df,file="Enriched_TF_RNA_seq_df.Rda")


significant_TFs_subsetted_UPe <- subset(pval_mat_UPe_df, pval_mat_UP_df$adj.p.value < 0.01)
enriched_TFs_UPe <- significant_TFs_subsetted_UPe$TF

summary(enriched_TFs_UPe)
enriched_TFs_UPe_fct <- as.factor(enriched_TFs_UPe)

distinct_enriched_TFs_UPe <- levels(enriched_TFs_UPe_fct)
length(distinct_enriched_TFs_UPe)
# 247
# There are 247 disticnt statistically overrepresented TFs

# summary(distinct_enriched_TFs_UPe == distinct_enriched_TFs)
# Mode    TRUE 
# logical     247

# plotting the outcomes
TF_ranking <- rankTFs( pval_mat_UPe, rankMethod = "gsea", makePlot = TRUE )
TF_ranking[[ "TFranking_plot" ]]

TF_enrichment_plot <- TF_ranking[[ "TFranking_plot" ]]

# saving the plot
pdf("TF_enrichment_plot_RNAseq.pdf")
TF_enrichment_plot
dev.off()


head( TF_ranking[[ "TF_ranking" ]] )
# TF       ES arg.ES pVal numberOfChIPs
# EZH2     EZH2  0.86527     21 0.00            17
# ZZZ3     ZZZ3  0.98582     17 0.00             2
# SUZ12   SUZ12  0.43264    604 0.78             6
# ZNF274 ZNF274  0.79905    217 0.00             5
# RAD21   RAD21  0.66571    367 0.01            19
# CBX3     CBX3 -0.57742    860 0.46             3

TF_ranking_df <- as.data.frame(TF_ranking$TF_ranking)
# Subsetting the TFs that surpass the 0.05 or 0.01 p-value threshold
TF_ranking_df_subset <- subset(TF_ranking_df, TF_ranking_df$pVal <= 0.01)

dim(TF_ranking_df_subset)
#  9 5
# 9 TFs are significantly upregulated
TF_ranking_df_subset

#  TF       ES arg.ES pVal numberOfChIPs
# EZH2   EZH2  0.96747     26 0.00            17
# KDM5B KDM5B -0.95843   1030 0.00             2
# MYC     MYC -0.68964    896 0.01            22
# YY1     YY1 -0.71297    781 0.01            17
# SIN3A SIN3A -0.74559    850 0.01            14
# NRF1   NRF1 -0.82173    862 0.00            11
# TAF1   TAF1 -0.88623    973 0.00            13
# MXI1   MXI1 -0.86407    909 0.00             8
# HCFC1 HCFC1 -0.89289    942 0.00             5

save(TF_ranking_df_subset,file="Sig_upregulated_TF_RNA_seq_df.Rda")


volcano <- plot_CM( pval_mat_UPe )

png("TF_enrichment_volcano_plot_RNAseq.png")
volcano
dev.off()

pdf("TF_enrichment_volcano_plot_RNAseq.pdf")
volcano
dev.off()

# Colouring the 9 TFs that were significanly upregulated in the treatment group
enriched_TFs <- c( "EZH2","KDM5B","MYC", "YY1", "SIN3A", "NRF1", "TAF1", "MXI1", "HCFC1" )

names(enriched_TFs) <- c( "EZH2","KDM5B","MYC", "YY1", "SIN3A", "NRF1", "TAF1", "MXI1", "HCFC1" )

colours()
# 657 colours in total
# red, lightblue, blue, darkgreen, yellow, purple, orange, pink, lightgreen, 
floor(runif(9, min=1, max=657))
# colour indices: 75 133 557 269 537 168 578 340 536


col <- c("red", "lightblue", "blue", "darkgreen", "yellow", "purple", "orange", "pink", "lightgreen")

plot_CM( pval_mat_UPe, specialTF = enriched_TFs, TF_colors = col ) #plot p-values against ORs highlighting indicated TFs

coloured_volcano <- plot_CM( pval_mat_UPe, specialTF = enriched_TFs, TF_colors = col) 


pdf("TF_enrichment_volcano_plot_RNAseq_coloured.pdf")
coloured_volcano
dev.off()

png("TF_enrichment_volcano_plot_RNAseq_coloured.png")
coloured_volcano
dev.off()
